### PR TITLE
Add ntfszapfree to optimize sparse images

### PR DESCRIPTION
**Pull request by felfert**
This PR adds an invocation of [ntfszapfree](https://github.com/felfert/ntfszapfree) in the final cleanup stage just before running sysprep. zapfree writes zeroes to the unused parts of the disk. This in turn helps qemu-img to drastically reduce the size of the resulting qcow2 image.

In order to see the effect, **after** the initial run of the qcow2 image in a hypervisor (which inflates the image quite extensive due to running windows-update) you now can run an additional convert in qemu-img like this:

`qemu-img convert -O qcow2 w2k16-cloudbase-init.qcow2 w2k16-cloudbase-init2.qcow2`

The result is quite impressive. Here are the sizes before and after running that additional convert:
```
-rw-r--r--. 1 root root 11G Jun  2 23:18 w2k16-cloudbase-init2.qcow2
-rw-r--r--. 1 root root 29G Jun  2 23:14 w2k16-cloudbase-init.qcow2
```
These are uncompressed images. With compression enabled this gets down to ~5G!

Cheers
 -Fritz

### CODE REVIEW

**Comment by ader1990** [comment]
Hello,

I also had a try at this: https://github.com/cloudbase/windows-openstack-imaging-tools/pull/140/files
but closed the PR due to external dependency, which I would like to avoid (sdelete or dd).

There is another way to very easily achieve the same thing using PowerShell:
https://sqlserverpowershell.com/2015/01/14/create-a-zerod-out-file-with-powershell/
It would be great if you could give it a try.

Thank you,
Adrian Vladu

**Comment by ader1990** [comment]
On another hand, to achieve the optimized, are you running with purge_updates=True, I suppose?
There is also the defrag/trim optimizations done: https://github.com/cloudbase/windows-openstack-imaging-tools/blob/master/WinImageBuilder.psm1#L939

That would be impressive if the sdelete will indirectly optimize the image and if we could find a native way (PS) to achieve it.

**Comment by felfert** [comment]
What is the reason why you don't want an external dependency? There is no copyright problem, because it is not included but downloaded at runtime.
IMHO, implementing it in PS/C# is "reinventing the wheel" and error-prone and therefore a bad idea.

-Fritz

**Comment by ader1990** [comment]
> What is the reason why you don't want an external dependency? There is no copyright problem, because it is not included but downloaded at runtime.
> IMHO, implementing it in PS/C# is "reinventing the wheel" and error-prone and therefore a bad idea.
> 
> -Fritz

Fair enough, but what I am trying here is to avoid an extra mandatory dependency problem.

I will make a review on the patch in order to avoid this issue, and not oblige users to download the tool (people should be able to run the code offline if they do not want to install updates) and make the sdelete run configurable and turned off by default.

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Examples/config.ini
+*.swp
```
Comment (ader1990) [inline .gitignore]: This PR contains too many things done at once in one commit.

My suggestions:
  - implement sdelete in one PR
  - gitignore, examples and other fixes should go in a separate PR

File: UnattendResources/Logon.ps1 [hunk 2]
```diff
@@ -75,9 +75,16 @@ function Clean-UpdateResources {
     Remove-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon" -Name AutoLogonCount -ErrorAction SilentlyContinue
 
     # Cleanup
-    Remove-Item -Recurse -Force $resourcesDir
+    Remove-Item -Recurse -Force -Exclude "sdelete.exe" $resourcesDir
     Remove-Item -Force "$ENV:SystemDrive\Unattend.xml"
-
+    $sdelete = "$resourcesDir\sdelete.exe"
```
Comment (ader1990) [inline UnattendResources/Logon.ps1]: Let s create a config.psm1 flag for sdelete enable.

I suggest to add sdelete_cleanup=False as default, grouped in section DEFAULT  in the config.psm1 and make the subsequent checks in the code in the logon.ps1 and .psm1 code.

File: UnattendResources/Logon.ps1 [hunk 3]
```diff
@@ -75,9 +75,16 @@ function Clean-UpdateResources {
...
+    Remove-Item -Recurse -Force -Exclude "sdelete.exe" $resourcesDir
     Remove-Item -Force "$ENV:SystemDrive\Unattend.xml"
-
+    $sdelete = "$resourcesDir\sdelete.exe"
+    if ( Test-Path $sdelete ) {
+        Write-Host "Optimizing for sparse image..."
+        & $sdelete -AcceptEULA -z $ENV:SystemDrive
+        Remove-Item -Recurse -Force $resourcesDir
+    } else {
+        Write-Debug "No sdelete. Image not optimized for qcow2"
```
Comment (ader1990) [inline UnattendResources/Logon.ps1]: remove qcow2 reference.

File: UnattendResources/Logon.ps1 [hunk 4]
```diff
@@ -75,9 +75,16 @@ function Clean-UpdateResources {
     Remove-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon" -Name AutoLogonCount -ErrorAction SilentlyContinue
 
     # Cleanup
-    Remove-Item -Recurse -Force $resourcesDir
+    Remove-Item -Recurse -Force -Exclude "sdelete.exe" $resourcesDir
     Remove-Item -Force "$ENV:SystemDrive\Unattend.xml"
-
+    $sdelete = "$resourcesDir\sdelete.exe"
+    if ( Test-Path $sdelete ) {
```
Comment (ader1990) [inline UnattendResources/Logon.ps1]: move this check if the flag is enabled or not

File: UnattendResources/Logon.ps1 [hunk 5]
```diff
@@ -125,6 +132,7 @@ function Install-WindowsUpdates {
         "6.3" = @("KB2887595")
     }
     $excludedUpdates = $KBIdsBlacklist[$OSKernelVersion]
+    Write-Host "Searching for updates..."
```
Comment (ader1990) [inline UnattendResources/Logon.ps1]: remove the unrelated code changes from the sdelete PR

File: WinImageBuilder.psm1 [hunk 6]
```diff
@@ -363,6 +363,7 @@ function Convert-VirtualDisk {
         $compressParam = "-c"
     }
     Write-Log "Converting virtual disk image from $vhdPath to $outPath..."
+    Write-Debug "Executing: $scriptPath\bin\qemu-img.exe convert $compressParam -O $($format.ToLower()) $vhdPath $outPath"
```
Comment (ader1990) [inline WinImageBuilder.psm1]: ditto

File: WinImageBuilder.psm1 [hunk 7]
```diff
@@ -504,6 +505,31 @@ function Download-CloudbaseInit {
...
+        [string]$resourcesDir,
+        [Parameter(Mandatory=$true)]
+        [string]$osArch
+    )
+    $SDeletePath = "$resourcesDir\sdelete.exe"
+    $SDelete64Path = "$resourcesDir\sdelete64.exe"
+    $SDeleteZipPath = "$resourcesDir\sdelete.zip"
+    Write-Log "Downloading SDelete..."
+
+    $SDeleteUrl = "https://download.sysinternals.com/files/SDelete.zip"
```
Comment (ader1990) [inline WinImageBuilder.psm1]: we should add this as a config option in a next PR,  I can add it later.

File: WinImageBuilder.psm1 [hunk 8]
```diff
@@ -1328,6 +1354,7 @@ function New-WindowsCloudImage {
         Copy-Item $ConfigFilePath "$resourcesDir\config.ini"
         Set-WindowsWallpaper -WinDrive $winImagePath -WallpaperPath $windowsImageConfig.wallpaper_path `
             -WallpaperSolidColor $windowsImageConfig.wallpaper_solid_color
+        Download-SDelete $resourcesDir ([string]$image.ImageArchitecture)
```
Comment (ader1990) [inline WinImageBuilder.psm1]: check for config flag whether to perform the download

File: WinImageBuilder.psm1 [hunk 9]
```diff
@@ -1461,6 +1488,7 @@ function New-WindowsFromGoldenImage {
         Copy-Item $ConfigFilePath "$resourcesDir\config.ini"
         Set-WindowsWallpaper -WinDrive $driveLetterGold -WallpaperPath $windowsImageConfig.wallpaper_path `
             -WallpaperSolidColor $windowsImageConfig.wallpaper_solid_color
+        Download-SDelete $resourcesDir ([string]$image.ImageArchitecture)
```
Comment (ader1990) [inline WinImageBuilder.psm1]: ditto

**Comment by felfert** [comment]
Well, the code in UnattendResources/Logon.ps1already skips execution if sdelete.exe is not available.
This just leaves a config-option in WinImageBuilder.psm1 which is easy.
I will add a commit to this PR ...

File: WinImageBuilder.psm1 (hunk 9 above)
Comment (ader1990) [inline WinImageBuilder.psm1]: image variable does not exist, I think it should be imageInfo

File: WinImageBuilder.psm1 (hunk 8 above)
Comment (ader1990) [reply WinImageBuilder.psm1]: image variable does not exist, I think it should be imageInfo

File: WinImageBuilder.psm1 (hunk 9 above)
Comment (felfert) [reply WinImageBuilder.psm1]: Huh? Thats not true! Just one line below is the Download-Cloudinit call which uses exactly the same.

Comment (felfert) [reply WinImageBuilder.psm1]: Oops yes, thats in this Golden thingy, you are right

File: UnattendResources/Logon.ps1 (hunk 3 above)
Comment (felfert) [reply UnattendResources/Logon.ps1]: Ok

File: .gitignore (hunk 1 above)
Comment (felfert) [reply .gitignore]: Ok

File: WinImageBuilder.psm1 (hunk 9 above)
Comment (felfert) [reply WinImageBuilder.psm1]: Just to clarify:
In New-WindowsCloudImage, $image DOES exist
In New-WindowsFromGoldenImage, $image DOES NOT exist. Instead, $imageInfo.imageArchitecture must be used.

Comment (ader1990) [reply WinImageBuilder.psm1]: got it, you are correct ;)

File: UnattendResources/Logon.ps1 (hunk 2 above)
Comment (ader1990) [reply UnattendResources/Logon.ps1]: The only problem I see here is that the user could set sdelete_cleanup flag on true and for some reason (like a bug in the code) sdelete.exe is not present and the image is presented as a best effort, but sdelete has not been run.
But I can add the extra check in another patch, to make sure that sdelete will be run.

File: UnattendResources/Logon.ps1 [hunk 10]
```diff
@@ -75,9 +75,16 @@ function Clean-UpdateResources {
...
 
     # Cleanup
-    Remove-Item -Recurse -Force $resourcesDir
+    Remove-Item -Recurse -Force -Exclude "sdelete.exe" $resourcesDir
     Remove-Item -Force "$ENV:SystemDrive\Unattend.xml"
-
+    $sdelete = "$resourcesDir\sdelete.exe"
+    if ( Test-Path $sdelete ) {
+        Write-Host "Optimizing for sparse image..."
+        & $sdelete -AcceptEULA -z $ENV:SystemDrive
```
Comment (ader1990) [inline UnattendResources/Logon.ps1]: I think here it would be nice to have a LASTEXITCODE check maybe?

**Comment by felfert** [comment]
Ok, PR should now have all the requested changes

**Comment by ader1990** [comment]
I will give it a run to make sure it works on my env and it should be good to merge. Thank you for the contribution, much appreciated!

File: UnattendResources/Logon.ps1 (hunk 10 above)
Comment (felfert) [reply UnattendResources/Logon.ps1]: I'm not shure, if sdelete probably uses some special exit codes. Will check its source.

Comment (felfert) [reply UnattendResources/Logon.ps1]: Regarding sdelete.exe accidentially not existing:

As far as I understand, Logon.ps1 should not be stopped on error anyway. So the only option that makes sense is to print a warning.

**Comment by ader1990** [comment]
@felfert Found a small unit test issue, you need to add sdelete_cleanup = true here: https://github.com/felfert/windows-openstack-imaging-tools/blob/add_sdelete/Tests/fake-config.ini#L13

**Comment by felfert** [comment]
Ah yes of course :-)

File: UnattendResources/Logon.ps1 (hunk 10 above)
Comment (felfert) [reply UnattendResources/Logon.ps1]: I checked the source of sdelete and came to the conclusion that a LASTEXITCODE check is unnecessary, because:

1. sdelete prints its own error messages
2. we would only be able to print a warning anyway

Comment (ader1990) [reply UnattendResources/Logon.ps1]: Got it, will do some runs and make sure the sdelete cannot break the generation also on older versions like win7.

Comment (felfert) [reply UnattendResources/Logon.ps1]: sdelete itself is ancient. It should work on any old WIndows. Quote from the source:
```
// SDelete - Secure Delete
// Copyright (C) 1999 Mark Russinovich
// Systems Internals - http://www.sysinternals.com
```

**Comment by alexpilotti** [comment]
@felfert first, thanks for contributing this (and nice to see you again!).

The SysInternals' license limits a lot of things unfortunately, for example if forbids to
"transfer the software or this agreement to any third party".
This implies that we cannot download and run the software on behalf of the user w/o them accepting explicitly the license (I'll refrain from saying what I think of that license). 

IMO the idea is good, I'm actually surprised that unallocated areas are not zeroed, we just need to find another tool.

**Comment by alexpilotti** [comment]
To unblock this, let's expect the user to have sdelete already in the path and let's add a "zero_unused_volume_sectors" boolean option to control it.

We can then replace sdelete with a simple open source project that zeroes the unused parts of the volume (basically just FSCTL_GET_VOLUME_BITMAP plus writing the zeros).

File: WinImageBuilder.psm1 [hunk 11]
```diff
@@ -1328,6 +1353,9 @@ function New-WindowsCloudImage {
         Copy-Item $ConfigFilePath "$resourcesDir\config.ini"
         Set-WindowsWallpaper -WinDrive $winImagePath -WallpaperPath $windowsImageConfig.wallpaper_path `
             -WallpaperSolidColor $windowsImageConfig.wallpaper_solid_color
+        if ($windowsImageConfig.sdelete_cleanup) {
```
Comment (ader1990) [inline WinImageBuilder.psm1]: remove this part, as sdelete can be downloaded in the external script.

File: WinImageBuilder.psm1 [hunk 12]
```diff
@@ -1461,6 +1489,9 @@ function New-WindowsFromGoldenImage {
         Copy-Item $ConfigFilePath "$resourcesDir\config.ini"
         Set-WindowsWallpaper -WinDrive $driveLetterGold -WallpaperPath $windowsImageConfig.wallpaper_path `
             -WallpaperSolidColor $windowsImageConfig.wallpaper_solid_color
+        if ($windowsImageConfig.sdelete_cleanup) {
```
Comment (ader1990) [inline WinImageBuilder.psm1]: remove this part, as sdelete can be downloaded in the external script.

File: Tests/WinImageBuilder.Tests.ps1 [hunk 13]
```diff
@@ -101,6 +101,7 @@ Describe "Test New-WindowsCloudImage" {
     Mock Copy-CustomResources -Verifiable -ModuleName $moduleName { return 0 }
     Mock Copy-Item -Verifiable -ModuleName $moduleName { return 0 }
     Mock Download-CloudbaseInit -Verifiable -ModuleName $moduleName { return 0 }
+    Mock Download-SDelete -Verifiable -ModuleName $moduleName { return 0 }
```
Comment (ader1990) [inline Tests/WinImageBuilder.Tests.ps1]: remove the changes to this file

File: Tests/fake-config.ini [hunk 14]
```diff
@@ -10,6 +10,7 @@ force = false
 zip_password = ""
 install_maas_hooks = false
 compression_format = tar.gz
+sdelete_cleanup = true
```
Comment (ader1990) [inline Tests/fake-config.ini]: ditto

File: UnattendResources/Logon.ps1 (hunk 10 above)
Comment (ader1990) [reply UnattendResources/Logon.ps1]: as @alexpilotti said, this part can be left as it is. if sdelete is present, it will use it.

File: WinImageBuilder.psm1 [hunk 15]
```diff
@@ -504,6 +504,31 @@ function Download-CloudbaseInit {
     }
 }
 
+function Download-SDelete {
```
Comment (ader1990) [inline WinImageBuilder.psm1]: this part can be moved into the user private image generation scripts (similar to the examples). this way, there won't be license issue.

File: Config.psm1 [hunk 16]
```diff
@@ -151,6 +151,10 @@ function Get-AvailableConfigOptions {
           "Description" = "If set to true and the target image format is QCOW2, the image conversion will
                            use qemu-img built-in compression. The compressed qcow2 image will be smaller, but the conversion
                            will take longer time."},
+        @{"Name" = "sdelete_cleanup"; "DefaultValue" = $false; "AsBoolean" = $true;
```
Comment (ader1990) [inline Config.psm1]: you can change this config name and description according to @alexpilotti suggestion. In the future, sdelete functionality can be replaced with n existing windows tool or an open source one.

**Comment by alexpilotti** [comment]
One more thing to evaliuate: a simple call to Optimize-VHD after the updates, before the conversion.

**Comment by felfert** [comment]
Yes. But I'm travelling right now. Will try on Friday

Am 4. Juni 2019 16:15:45 MESZ schrieb Adrian Vladu <notifications@github.com>:
>@felfert I created a patch that also runs Optimize-VHD on the image:
>https://github.com/cloudbase/windows-openstack-imaging-tools/pull/265.
>Can you give it a try and check if it solves the qcow2 issue?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/cloudbase/windows-openstack-imaging-tools/pull/261#issuecomment-498690659

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by felfert** [comment]
> To unblock this, let's expect the user to have sdelete already in the path and let's add a "zero_unused_volume_sectors" boolean option to control it.
> 
> We can then replace sdelete with a simple open source project that zeroes the unused parts of the volume (basically just FSCTL_GET_VOLUME_BITMAP plus writing the zeros).

I now have created an [open-source replacement](https://github.com/felfert/ntfszapfree) for sdelete and replaced all references of sdelete with this new tool named zapfree. As requested, the config switch was renamed to  **zero_unused_volume_sectors**

**Comment by felfert** [comment]
@ader1990 @alexpilotti 
Did you see my commits, replacing of sdelete with open-source ntfszapfree?
This should unblock this PR

**Comment by alexpilotti** [comment]
@felfert thanks for creating the seekers replacement, we’re currently doing some tests, as soon as they finish we can merge!

File: WinImageBuilder.psm1 [hunk 17]
```diff
@@ -504,6 +504,31 @@ function Download-CloudbaseInit {
...
+        [Parameter(Mandatory=$true)]
+        [string]$resourcesDir,
+        [Parameter(Mandatory=$true)]
+        [string]$osArch
+    )
+    $ZapFreePath = "$resourcesDir\zapfree.exe"
+    $ZapFree32Path = "$resourcesDir\zapfree32.exe"
+    $ZapFreeZipPath = "$resourcesDir\ntfszapfree.zip"
+    Write-Log "Downloading ntfszapfree..."
+
```
Comment (ader1990) [inline WinImageBuilder.psm1]: @felfert you are required to add here a secure tls connection setting
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12

Otherwise, the download will fail with this error:

```
6/14/2019 1:30:46 PM - Downloading ntfszapfree...
WARNING: Exception calling "DownloadFile" with "2" argument(s): "The request was aborted: Could not create SSL/TLS secure channel."
WARNING: Exception calling "DownloadFile" with "2" argument(s): "The request was aborted: Could not create SSL/TLS secure channel."
WARNING: Exception calling "DownloadFile" with "2" argument(s): "The request was aborted: Could not create SSL/TLS secure channel."
6/14/2019 1:30:59 PM - Exception calling "DownloadFile" with "2" argument(s): "The request was aborted: Could not create SSL/TLS secure channel."
```

Comment (felfert) [reply WinImageBuilder.psm1]: Hmm strange. Why does Download-CloudbaesInit not require this as well?
Perhaps your download server still offers Tls1?
Anyway: done

Comment (ader1990) [reply WinImageBuilder.psm1]: I have added a separate PR for the TLS enforcement: https://github.com/cloudbase/windows-openstack-imaging-tools/pull/273/files that enables Tls1.2 globally.

GitHub now requires TLS1.2, while cloudbase.it or fedora virtio repo do not.

Comment (felfert) [reply WinImageBuilder.psm1]: Ok, so I reverted my last commit

File: UnattendResources/Logon.ps1 [hunk 18]
```diff
@@ -75,9 +75,16 @@ function Clean-UpdateResources {
     Remove-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon" -Name AutoLogonCount -ErrorAction SilentlyContinue
 
     # Cleanup
-    Remove-Item -Recurse -Force $resourcesDir
+    Remove-Item -Recurse -Force -Exclude "zapfree.exe" $resourcesDir
     Remove-Item -Force "$ENV:SystemDrive\Unattend.xml"
-
```
Comment (ader1990) [inline UnattendResources/Logon.ps1]: I think the placement of zapfree execution should not be here, as it is optimistically at best.
At this stage, sysprep is already executing and sysprep might finish and shutdown the VM before zapfree finishes, leaving a possible broken image or a huge file created by zapfree.

Maybe https://github.com/cloudbase/windows-openstack-imaging-tools/pull/261/files#diff-b0340988ad4db4e9c00077d542df3922R401 (putting zapfree right before the sysprep) makes more sense.

Comment (ader1990) [reply UnattendResources/Logon.ps1]: To give more insight, sysprep is executed async mode, and it will turn off the vm after it finishes.

Comment (felfert) [reply UnattendResources/Logon.ps1]: This raises the question:
Why are there any actions *after* sysprep at all if they are not guaranteed to finish?

Comment (ader1990) [reply UnattendResources/Logon.ps1]: A sysprep basic run takes around several minutes, while the unattend resources cleanup takes a few seconds. But you have a point there. Is a little of cat and mouse problem, as I cannot remove the unattend resource dir before as it contains the unattend.xml file or some Run-CustomScript "RunAfterSysprep.ps1" hook required files. The best way is to remove the /shutdown flag to sysprep, then add a wait loop until syspre.exe finishes (by checking that the process exists) and then perform cleanup/zap + the manual shutdown.

Comment (felfert) [reply UnattendResources/Logon.ps1]: Well, that's a possible topic for a new PR. For now, I just moved the zap invocation before the sysprep invocation.

**Comment by ader1990** [comment]
Looks good, execution tested on a 2016 golden image:

```
6/14/2019 4:03:35 PM - Cloud image from golden image generation started.
6/14/2019 4:03:46 PM - Copy Unattend Resources: Windows Server 2012 R2 SERVERSTANDARD...
6/14/2019 4:03:46 PM - Copying: C:\Users\windows-openstack-imaging-tools\UnattendResources G:\UnattendResources
6/14/2019 4:03:46 PM - Resources have been copied.
6/14/2019 4:03:46 PM - Copy Custom Resources: ...
6/14/2019 4:03:46 PM - Custom Resources at: G:\UnattendResources.
6/14/2019 4:03:46 PM - Setting wallpaper...
6/14/2019 4:03:46 PM - Wallpaper copied to the image.
6/14/2019 4:03:46 PM - Cached wallpaper for user Administrator has been replaced.
6/14/2019 4:03:46 PM - Wallpaper GPO copied to the image.
6/14/2019 4:03:46 PM - Wallpaper was set.
6/14/2019 4:03:46 PM - Downloading ntfszapfree...
6/14/2019 4:03:48 PM - Downloading Cloudbase-Init...
6/14/2019 4:04:07 PM - Waiting for WindowsGoldImage-Sysprep323236298 to finish sysprep.
6/14/2019 4:08:23 PM - Shrinking VHD to minimum size
6/14/2019 4:08:23 PM - Initial VHD size is: 30 GB
6/14/2019 4:08:30 PM - Current partition size: 29.998046875 GB
6/14/2019 4:08:30 PM - New partition size: 13 GB
6/14/2019 4:08:30 PM - Size increased: 13958643712
WARNING: Size Not Supported

Extended information:
The specified shrink size is too big and will cause the volume to be smaller than the
minimum volume size.


Activity ID: {4ba4fabb-7c02-4d59-9951-7a9a156a8d94}
6/14/2019 4:08:40 PM - Size increased: 14063501312
6/14/2019 4:08:52 PM - Final disk size: 13.099609375 GB
6/14/2019 4:08:52 PM - Optimize VHD E:\diskimages\zap.vhdx: file size before optimization is 23.12890625 GB
6/14/2019 4:09:06 PM - Optimize VHD E:\diskimages\zap.vhdx: file size after optimization is 12.03515625 GB
6/14/2019 4:09:06 PM - Cloud image from golden image generation finished.
```

**Comment by ader1990** [comment]
@felfert  Thank you for the contribution, will add what we discussed in the comment sections as issues to improve on, like the sysprep wait and proper cleanup,
